samza-core/src/test/java/org/apache/samza/zk/TestZkUtils.java

```diff
@@ -463,4 +459,44 @@ public void testgetNextJobModelVersion() {
     // Get on the JobModel version should return 2, taking into account the published version 2.
     Assert.assertEquals("3", zkUtils.getNextJobModelVersion(zkUtils.getJobModelVersion()));
   }
+
+
+  @Test
+  public void testCloseShouldRetryOnceOnInterruptedException() {
+    ZkClient zkClient = Mockito.mock(ZkClient.class);
+    ZkUtils zkUtils = new ZkUtils(KEY_BUILDER, zkClient,
+        SESSION_TIMEOUT_MS, new NoOpMetricsRegistry());
+
+    Mockito.doThrow(new ZkInterruptedException(new InterruptedException()))
+           .doAnswer(invocation -> null)
+           .when(zkClient).close();
+
+    zkUtils.close();
+
+    Mockito.verify(zkClient, Mockito.times(2)).close();
+  }
+
+  @Test
+  public void testCloseShouldTearDownZkConnectionOnInterruptedException() throws Exception {
+    CountDownLatch latch = new CountDownLatch(1);
+    ZkClient zkClient = new ZkClient("127.0.0.1:" + zkServer.getPort());
+    ZkUtils zkUtils = new ZkUtils(KEY_BUILDER, zkClient, SESSION_TIMEOUT_MS, new NoOpMetricsRegistry());
+
+    Thread threadToInterrupt = new Thread(() -> {
+        try {
+          latch.await();
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+        }
+        zkUtils.close();
+      });
+
+    threadToInterrupt.start();
+    threadToInterrupt.interrupt();
+    threadToInterrupt.join();
+
```

Comment:
We assert that the client connection is closed .. Should we assert that it was "connected" prior to invoking interrupt()?

Reply:
@vjagadish1989 
`ZkClient`  setups connection with the zookeeper server during instantiation. Here's the ZkClient constructor:
```
    public ZkClient(final IZkConnection zkConnection, final int connectionTimeout, final ZkSerializer zkSerializer, final long operationRetryTimeout) {
        if (zkConnection == null) {
            throw new NullPointerException("Zookeeper connection is null!");
        }
        _connection = zkConnection;
        _zkSerializer = zkSerializer;
        _operationRetryTimeoutInMillis = operationRetryTimeout;
        _isZkSaslEnabled = isZkSaslEnabled();
        connect(connectionTimeout, this);
    }
```
`ZkClient` constructor used in the test invoke this through a call chain. 

Added a assertion before interruption that the connection wasn't closed. There're no other state exposed from `ZkConnection` interface, hence this is only possible check we can do.